### ISSUE TITLE
Support the union of types for a property in model definition in CADL

### ISSUE THREAD

**Issue by lirenhe** [Azure/autorest.java#1891]
Latest design https://github.com/Azure/autorest.java/issues/1891#issuecomment-1381285183

---

in https://github.com/Azure/azure-rest-api-specs/blob/feature/cognitiveservices/openai-cadl/specification/cognitiveservices/OpenAI.Inference/models/embeddings.create.cadl#L22, we have the model defined as

input: string | string[] | integer[] | integer[][]; in the follow model.

```
alias EmbeddingsRequest = {
    @doc("The ID of the end-user, for use in tracking and rate-limiting.")
    user?: string;

    @doc("input type of embedding search to use")
    input_type?: string;

    @doc("ID of the model to use")
    "model"?: string;

    @doc("""
    An input to embed, encoded as a string, a list of strings, or a list of token
    lists
    """)
    input: string | string[] | integer[] | integer[][];
};
```
We need to support union of types for a property in model definition in CADL

This concept is similar to AnyOf in openapi 3 so I believe this is a valid case.

**Comment by weidongxu-microsoft** [comment]
anyOf is also not supported

**Comment by weidongxu-microsoft** [comment]
In this case of Union in request, a possible model would be

```java
EmbeddingsRequest setInput(String);

EmbeddingsRequest setInput(List<String>);

EmbeddingsRequest setInput(List<Integer>);

EmbeddingsRequest setInput(List<List<Integer>>);

Object getInput();
```

However, the `setInput(List<Integer>)` already conflict with `setInput(List<List<Integer>>)`, as Java erases the generic.

---

Union likely won't be supported if it is in response.

---

At present, Union is generated as Object (Unknown).

**Comment by johanste** [comment]
My question is what you would want the API to look like if it was written by hand?

**Comment by weidongxu-microsoft** [comment]
Another way maybe a Generic like this https://github.com/pakoito/JavaSealedUnions

This might solve the general problem better, but likely not very easy to use.
Also given the Jackson, likely it and the related Jackson classes need to be in core.

**Comment by weidongxu-microsoft** [comment]
For overload:

If Union appears on API signature, one might prefer to move it to option bag.

- It is simpler to overload a single setter in option bag (aka model), than overload the API multiple times
- If there is multiple Union on API signature, the overload could cause combinatorial explosion

**Comment by weidongxu-microsoft** [comment]
Currently solution

```java
class InputModel {
   private inputString;
   private inputStringArray;
   private inputIntegerArray;
   private inputIntegerArrayArray;

   // getter setter *4
}
```

There be corresponding code in convenience API, e.g.
```java
    input = inputModel.getInputString();
    if (input == null) {
        input = inputModel.getInputStringArray();
    }
    ...
```

**Comment by srnagar** [comment]
Minor update:

```java
public class InputModel {
   private String inputString;
   private List<String> inputStringList;
   private List<Integer> inputIntegerList;
   private List<List<Integer>> inputListOfIntegerList; // naming conventions may change but should use List instead of Array

   // getter setter *4
   // setter methods will not be fluent and will return void
   // class will not be annotation with @Fluent 
   // and we may likely introduce another annotation @Union in azure-core to represent this type
}
```

**Comment by weidongxu-microsoft** [comment]
I am trying to make the Union class serializable by Jackson via `JsonValue`. (stream would need another impl)

This way, we can mostly remove the code in API. And it would make life much simpler, when the Union exists in nested model (say 3 level deep).

Currently I haven't figured out how to throw exception there, when `getValue()` found the class get multiple values set.

```java
/** The InputModel model. */
public final class InputModel {
    private String inputString;

    private List<String> inputStringArray;

    private List<Integer> inputIntegerArray;

    private List<List<Integer>> inputIntegerArrayArray;

    /** Creates an instance of InputModel class. */
    public InputModel() {}

    @JsonValue
    private Object getValue() {
        Object value = null;
        if (value == null) {
            value = this.inputString;
        }
        if (value == null) {
            value = this.inputStringArray;
        }
        if (value == null) {
            value = this.inputIntegerArray;
        }
        if (value == null) {
            value = this.inputIntegerArrayArray;
        }
        return value;
    }

    /**
     * Get the inputString property: The inputString property.
     *
     * @return the inputString value.
     */
    public String getInputString() {
        return this.inputString;
    }

    /**
     * Set the inputString property: The inputString property.
     *
     * @param inputString the inputString value to set.
     */
    public void setInputString(String inputString) {
        this.inputString = inputString;
    }

    /**
     * Get the inputStringArray property: The inputStringArray property.
     *
     * @return the inputStringArray value.
     */
    public List<String> getInputStringArray() {
        return this.inputStringArray;
    }

    /**
     * Set the inputStringArray property: The inputStringArray property.
     *
     * @param inputStringArray the inputStringArray value to set.
     */
    public void setInputStringArray(List<String> inputStringArray) {
        this.inputStringArray = inputStringArray;
    }

    /**
     * Get the inputIntegerArray property: The inputIntegerArray property.
     *
     * @return the inputIntegerArray value.
     */
    public List<Integer> getInputIntegerArray() {
        return this.inputIntegerArray;
    }

    /**
     * Set the inputIntegerArray property: The inputIntegerArray property.
     *
     * @param inputIntegerArray the inputIntegerArray value to set.
     */
    public void setInputIntegerArray(List<Integer> inputIntegerArray) {
        this.inputIntegerArray = inputIntegerArray;
    }

    /**
     * Get the inputIntegerArrayArray property: The inputIntegerArrayArray property.
     *
     * @return the inputIntegerArrayArray value.
     */
    public List<List<Integer>> getInputIntegerArrayArray() {
        return this.inputIntegerArrayArray;
    }

    /**
     * Set the inputIntegerArrayArray property: The inputIntegerArrayArray property.
     *
     * @param inputIntegerArrayArray the inputIntegerArrayArray value to set.
     */
    public void setInputIntegerArrayArray(List<List<Integer>> inputIntegerArrayArray) {
        this.inputIntegerArrayArray = inputIntegerArrayArray;
    }
}
```

**Comment by weidongxu-microsoft** [comment]
> Minor update:
> 
> ```java
> public class InputModel {
>    private String inputString;
>    private List<String> inputStringList;
>    private List<Integer> inputIntegerList;
>    private List<List<Integer>> inputListOfIntegerList; // naming conventions may change but should use List instead of Array
> 
>    // getter setter *4
>    // setter methods will not be fluent and will return void
>    // class will not be annotation with @Fluent 
>    // and we may likely introduce another annotation @Union in azure-core to represent this type
> }
> ```

@srnagar 

When writing the test, I find non-fluent could have some negative impact on code.

I cannot write it this way.

```java
        SendLongOptions options = new SendLongOptions("id", "input", 1)
                .setDataUnion(new DataUnionModel().setDataUnionString("dataString"));
```
or this way
```java
        DataUnionModel dataUnion = new DataUnionModel().setDataUnionString("dataString"));
```

So we might still weigh on which way to go.
Anyway this would be minor change to codegen.

**Comment by weidongxu-microsoft** [comment]
Latest design

`input: string | string[]` will generate something like

```java
public abstract class InputModel {
}

public StringInputModel extends InputModel {
    private String inputString;

    public String getInputString();

    public StringInputModel setInputString(String inputString);
}

public StringListInputModel extends InputModel {
    private List<String> inputStringList;

    public List<String> getInputStringList();

    public StringInputModel setInputString(List<String> inputStringList);
}
```

---

There could be a `InputTypeExpandableEnum type;` and `public InputTypeExpandableEnum getType();` in superclass and get overloaded in subclasses. Currently it is not in Jan. plan.

**Comment by srnagar** [comment]
Also, codegen will support union types straight from service cadl by creating the models similar to the one above. The model names will be reviewed and most likely renamed through a sidecar. We'll not make sidecar mandatory in codegen when a cadl has union types so, we can get an initial version of the client library generated for review.

**Comment by m-nash** [comment]
@weidongxu-microsoft any reason you don't go with a discriminator pattern here where your base has a `kind` property?

**Comment by weidongxu-microsoft** [comment]
@m-nash

I've mentioned here

> There could be a InputTypeExpandableEnum type; and public InputTypeExpandableEnum getType(); in superclass and get overridden in all subclasses. Currently it is not in Jan. plan (as we got one more class name to deduce and the type might not be very helpful).

We can add it if we find it useful for Java user.

The fact is that at present, even when service does have real discriminator, Java still does not expose it to user (it reflected in polymorphism structure).
Example: https://github.com/Azure/autorest.java/blob/main/vanilla-tests/src/main/java/fixtures/bodycomplex/models/Fish.java
You can see we don't have `fishtype` in setter/getter or variable.

So in this case of the fictional discriminator (there is no "kind" or "type" on wire), I'd personally want to hold this back for a while.

---

In latest Java 17, user can do pattern matching without any switch on "kind" or "type" (if they switch on the "kind" discriminator  they still need to case before using the object, the "kind" thing is not really a Java thing)
```java
record Point(int x, int y) { }
enum Color { RED, GREEN, BLUE; }
...
    static void typeTester(Object obj) {
        switch (obj) {
            case null     -> System.out.println("null");
            case String s -> System.out.println("String");
            case Color c  -> System.out.println("Color with " + c.values().length + " values");
            case Point p  -> System.out.println("Record class: " + p.toString());
            case int[] ia -> System.out.println("Array of int values of length" + ia.length);
            default       -> System.out.println("Something else");
        }
    }
```
https://docs.oracle.com/en/java/javase/18/language/pattern-matching-switch-expressions-and-statements.html

**Comment by weidongxu-microsoft** [comment]
> Why not simply generate as a generic "object" -- i.e. an "any" type (what Cadl calls "unknown").
> 
> How does the Java generator handle a Cadl property defined as "unknown"?

Yes, `Object` is the first mitigate solution before we move here, that Cadl `Union` is just treated as `Unknown`.
https://apiview.dev/Assemblies/Review/cff7890772ed40e8a5785e5efda9d2e2?revisionId=d7d5836fd9744f21b6d29965f9f8426f&doc=False

But the apparent down side is that it is very difficult for user to know what object can be allowed and what cannot be allowed. User can only read the Javadoc and find what type is allowed.

**Comment by mikekistler** [comment]
I agree that having something more specific than "object" makes for a better user experience, but "object" is the best we would be able to do in DPG 1.0 since it did not support OpenAPI 3 features like "oneOf".

My concern here is treating this as a requirement for DPG 2.0. We have a solution that works at least for the case needed by OpenAI. General support for unions (i.e. full support for OpenAPI 3 "oneOf") will be a big task that I doubt we could contain in our schedule. Can we just go with the simple solution for now and drop this from the DPG 2.0 requirements?

**Comment by weidongxu-microsoft** [comment]
> I agree that having something more specific than "object" makes for a better user experience, but "object" is the best we would be able to do in DPG 1.0 since it did not support OpenAPI 3 features like "oneOf".
> 
> My concern here is treating this as a requirement for DPG 2.0. We have a solution that works at least for the case needed by OpenAI. General support for unions (i.e. full support for OpenAPI 3 "oneOf") will be a big task that I doubt we could contain in our schedule. Can we just go with the simple solution for now and drop this from the DPG 2.0 requirements?

DPG (without Cadl) won't have any model generated, and hence there is extremely unlikely we see any "oneOf" (either from OpenAPI 3 or some new "x-ms-one-of" extension) that need to be generated to code.
Hence, I won't worry about whether we need to backport it to DPG without Cadl.

The feature in cadl-java (at least for input/request) is shipped in 0.2.2 release.

I am fine if this is not needed and we decided to fallback to initial `Object`. But then this issue should not be created in the first place... (the `Object` was the Java solution before this issue)

**Comment by weidongxu-microsoft** [comment]
Pending:

- Support customization for these subclasses (name etc.) from cadl-dpg https://github.com/Azure/autorest.java/issues/1891#issuecomment-1381325018
- Design/impl when Union is used in output
